Add Duffing spring example

eh111eh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Pull Request #26627
Issue #26625

#### Brief description of what is fixed or changed

Add a system consisting of a Duffing oscillator with a pendulum.
<img width="428" alt="Screenshot 2024-06-11 at 19 27 56" src="https://github.com/sympy/sympy/assets/97640870/2a9e90c6-4a21-449d-9b18-b0a7659f57ef">

#### Other comments

Reference: P. Brzeskia, P. Perlikowskia, S. Yanchukb, T. Kapitaniaka, The dynamics of the pendulum suspended on the forced Duffing oscillator, Journal of Sound and Vibration, 2012, https://doi.org/10.48550/arXiv.1202.5937.

@tjstienstra 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

eh111eh committed:
added duffingspring actutator

eh111eh committed:
add a blank line

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

moorepants commented on doc/src/tutorials/duffing-example.rst:
```diff
@@ -0,0 +1,258 @@
+   ...     sign_correction = np.sign(x_val)
+   ...     force = force_func(x_val, alpha_val, beta_val)
+   ...     corrected_force = sign_correction * force
+   ...     return corrected_force
```
If this is an error in the `DuffingSpring`, then it should be corrected in a new PR (especially before SymPy 1.13 is released).

eh111eh replied:
I double checked the actuator and found no errors related to the sign. What I initially thought was an error actually arose from my misunderstanding of how the actuator handles the reference frame and the model. I can confirm that the restoring force is functioning properly. Based on the resources you suggested, I am now considering focusing on building a multibody system with this actuator.

moorepants commented on doc/src/tutorials/duffing-example.rst:
```diff
@@ -0,0 +1,258 @@
+   >>> # Extract the time and displacement from the solution
+   >>> time = solution.t
+   >>> displacements = solution.y[0]
+   >>> velocities = solution.y[1]
```
You are solving the differential equations independently of the formulation of the equations of motion with SymPy mechanics. The goal should be to demonstrate how you include the DuffingSpring in a multibody dynamics model as a general force actuator and then simulate the resulting equations of motion from that model. I recommend studying this example carefully: https://docs.sympy.org/dev/tutorials/biomechanics/biomechanical-model-example.html and see how we obtain the model's equations of motion with `KanesMethod` and then pass those to SciPy's integrator. You could even swap the muscle actuator out for your Duffing spring to test it in that arm model.

You also likely need to, in general, see how to make multibody models from these resources:

https://docs.sympy.org/dev/modules/physics/mechanics/examples.html

https://pydy.readthedocs.io/en/latest/index.html#examples

https://moorepants.github.io/learn-multibody-dynamics/index.html

The simplest possible model is a single mass moving in a single direction with a Diffing spring attaching it to a wall and you can write the equations very quickly without SymPy mechanics (as you've done), but that isn't what we hope you present here. The purpose of Actuators are to easily add them into multibody models.

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

eh111eh committed:
Change duffing-example

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

eh111eh committed:
Update duffing-example.rst

Locally passed bin/doctest --force-colors

eh111eh committed:
Add schematic diagram

eh111eh committed:
Update duffing-example.rst

A = N since q1 represents the vertical position, not the rotation.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> import sympy.physics.mechanics as me
+   >>> from sympy.physics.mechanics.actuator import DuffingSpring
+   >>> from sympy.physics.mechanics.lagrange import LagrangesMethod
+   >>> from sympy.physics.mechanics.pathway import LinearPathway
```
`me.<thing>` should give access to `DuffingSpring`, `LagrangesMethod`, `LinearPathway`.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :context: close-figs
+   :nofigs:
+
+   >>> B.set_ang_vel(N, u2 * N.z)
```
```suggestion
   >>> B.set_ang_vel(N, q2.diff() * N.z)
```

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :context: close-figs
+   :nofigs:
+
+   >>> # Define Rayleigh dissipation
```
Use normal text not a comment in the code.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :context: close-figs
+   :nofigs:
+
+   >>> # Kinetic Energy
```
Normal text

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> KPendulum = (1/2) * m * Pendulum.vel(N).dot(Pendulum.vel(N))
+   >>> Kinetic = KBlock + KPendulum
+   >>> Kinetic
+   0.5*M*u1(t)**2 + 0.5*m*(l**2*u2(t)**2 - 2*l*u1(t)*u2(t)*sin(q2(t)) + u1(t)**2)
```
If you create `RigidBody` you can then use `RigidBody.kinetic_energy()`.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :nofigs:
+
+   >>> L = Kinetic - Potential
+   >>> L
```
Recommend using the `lagrangian()` function.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+References
+==========
+
+P. Brzeskia, P. Perlikowskia, S. Yanchukb, T. Kapitaniaka,
```
Use the restructuredtext citation format: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#citations

moorepants commented on doc/src/tutorials/physics/index.rst:
```diff
@@ -22,3 +22,4 @@ write to our `mailing list
    :maxdepth: 2
 
    biomechanics/index.rst
+   duffing-example.rst
```
This should be under `mechanics/duffing-example.rst`.

tjstienstra replied:
We should also create an `mechanics/index.rst` file, but this can be addressed when solving #25607

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> u1, u2 = me.dynamicsymbols('u1 u2')
+   >>> u1d, u2d = me.dynamicsymbols('u1 u2', 1)
+
+- :math:`q1`: Generalized coordinate representing the position of the Duffing oscillator
```
```suggestion
- :math:`q_1`: Generalized coordinate representing the position of the Duffing oscillator
```

use latex math in `:math:`

eh111eh committed:
Update doc/src/tutorials/physics/duffing-example.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :format: doctest
+   :include-source: True
+   :context: close-figs
+   :nofigs:
```
If you do not intend to show plots in this example, then it is better to use the normal doctest style:

```python
>>> import sympy as sm
>>> sm.symbols('x')
x
```

moorepants replied:
Without the plot directive.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+
+In this example we demonstrate the use of functionality provided in
+:obj:`sympy.physics.mechanics` for deriving the quations of motion (EOM) for a system
+consisting of a Duffing oscillator with a pendulum.
```
Make a citation to the paper here in the introduction.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+- :math:`u1`: Generalized speed associated with the Duffing oscillator
+- :math:`u2`: Generalized speed associated with the pendulum
+- :math:`u1d`: First time derivative of `u1`, representing the acceleration of the Duffing oscillator
+- :math:`u2d`: First time derivative of `u2`, representing the angular acceleration of the pendulum
```
```suggestion
- :math:`\dot{u}_2`: First time derivative of `u2`, representing the angular acceleration of the pendulum
```

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :context: close-figs
+   :nofigs:
+
+   >>> # Define reference frames
```
Move comments to normal text.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+
+   >>> # Define reference frames
+   >>> N = me.ReferenceFrame('N')
+   >>> A = N # Duffing Oscillator Block
```
Not needed. Only use new frames when there is orientation between frames.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> # Define reference frames
+   >>> N = me.ReferenceFrame('N')
+   >>> A = N # Duffing Oscillator Block
+   >>> B = N.orientnew('B', 'axis', (q2, N.z)) # Pendulum
```
Move comments to actual explanatory text.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   :context: close-figs
+   :nofigs:
+
+   >>> B.set_ang_vel(N, q2.diff() * N.z)
```
Show the result:

```python
B.ang_vel_in(N)
```

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+
+   >>> O = me.Point('O') # fixed point in the inertial reference frame
+   >>> Block = O.locatenew('Block', q1 * N.y)
+   >>> Pendulum = Block.locatenew('Pendulum', l * B.y)
```
Better to label points in the diagram and use those names for points. The pendulum has two points for example, so this is currently ambiguous.

moorepants replied:
You can save "block" and "pendulum" for `RigidBody` names.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> O.set_vel(N, 0)
+   >>> Block.set_vel(N, u1 * N.y)
+   >>> Pendulum.v2pt_theory(Block, N, B)
+   u1(t)*N.y - l*u2(t)*B.x
```
No u's should be present in Lagrange formulation.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   u1(t)*N.y - l*u2(t)*B.x
+
+   >>> ParBlock = me.Particle('ParBlock', Block, M)
+   >>> ParPendulum = me.Particle('ParPendulum', Pendulum, m)
```
Try to follow pep8 for variable names: `par_block`. In mechanics examples we don't often follow it when using single letter names for things, i.e. `A = ReferenceFrame('A')` so that is an exception to the rule.

moorepants commented on doc/src/tutorials/physics/duffing-example.rst:
```diff
@@ -0,0 +1,183 @@
+   >>> pathway = LinearPathway(O, Block)
+   >>> duffing_spring = DuffingSpring(k1, k2, pathway, 0)
+   >>> duffing_force = -duffing_spring.force
+   >>> duffing_force
```
The idea is to use the `to_loads()` method when you have an actuator. See the biomechanics examples.

eh111eh committed:
Update doc/src/tutorials/physics/duffing-example.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

eh111eh committed:
Update doc/src/tutorials/physics/duffing-example.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

eh111eh committed:
Update duffing-example.rst

All feedback applied, except `to_loads` method for DuffingSpring actuator. Still investigating `to_loads` method in biomechanics examples.

eh111eh committed:
Merge branch 'duffingexample2' of https://github.com/eh111eh/SymPy into duffingexample2

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,135 @@
+   >>> O.set_vel(N, 0)
+   >>> Block.set_vel(N, q1d * N.y)
+   >>> Pendulum.v2pt_theory(Block, N, B)
+   Derivative(q1(t), t)*N.y - l*Derivative(q2(t), t)*B.x
```
You can use `me.init_vprinting()` to have pretty printed output for these lines.

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,135 @@
+
+Define Rayleigh dissipation.
+
+   >>> D = (1/2) * c1 * q1d**2
```
For this term it would be ideal if we had a `LinearDamper()` acuator, then you can pass use `.to_loads()` from both to generate the force list for `LagrangesMethod`. If we don't have that yet, then it would still be best to add this a a load to `LagrangesMethod` instead of forming the generalized force manually.

moorepants replied:
Also, you write the energy term here as `D` but then you pass it in as a force below in the force list. `D` is not a force but the energy lost due to a linear damping force. Best to define as a force `-c1*q1d*N.y`.

eh111eh committed:
Update duffing-example.rst

Attempted `to_loads` method and me.init_vprinting(). I deleted manual calculation of potential and kinetic energies since `to_loads` does all necessary jobs for LagrangesMethod.

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,148 @@
+   ...
+   ...         self.pathway1 = me.LinearPathway(self.P1, self.P2)
+   ...         self.duffing_spring = me.DuffingSpring(self.k1, self.k2, self.pathway1, 0)
+   ...         self.pathway2 = me.LinearPathway(self.P2, self.P3)
```
Why would the pathways be different for the spring and the damper? Aren't they in parallel?

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,148 @@
+   ...         damper_force = -self.c1 * q1d * N.y
+   ...         total_force_block = duffing_force + damper_force
+   ...         loads = [(self.P2, total_force_block), (self.P3, -self.m * self.g * N.y)]
+   ...         return loads
```
Given that your GSoC is about adding actuators, why don't you just add a `LinearDamper`?

eh111eh replied:
Thank you for the advice, so then would it be like this?
```
class DuffingActuator(me.ActuatorBase):
    def __init__(self, P1, P2, linear_stiffness, nonlinear_stiffness, damping, pendulum_mass, block_mass, g):
        self.P1 = P1
        self.P2 = P2
        self.k1 = linear_stiffness
        self.k2 = nonlinear_stiffness
        self.c1 = damping
        self.m = pendulum_mass
        self.M = block_mass
        self.g = g

        self.pathway = me.LinearPathway(self.P1, self.P2)
        self.duffing_spring = me.DuffingSpring(self.k1, self.k2, self.pathway, 0)

    def to_loads(self):
        damper = me.LinearDamper(c1, self.pathway)
        damper.force

        duffing_force = self.duffing_spring.force * N.y

        loads = [(self.P1, duffing_force + damper.force * N.y), (self.P2, -self.m * self.g * N.y)]
        return loads
```

Or do we include only duffing force in the DuffingActuator class and separately obtain damping from `LinearDamper`?

eh111eh replied:
I've also thought of this way:
```
class DuffingActuator(me.ActuatorBase):
    def __init__(self, P1, P2, linear_stiffness, nonlinear_stiffness, damping, pendulum_mass, block_mass, g, equilibrium_length):
        self.P1 = P1
        self.P2 = P2
        self.k1 = linear_stiffness
        self.k2 = nonlinear_stiffness
        self.c1 = damping
        self.m = pendulum_mass
        self.M = block_mass
        self.g = g
        self.l = equilibrium_length

        self.pathway = me.LinearPathway(self.P1, self.P2)

    def to_loads(self):
        damper = me.LinearDamper(c1, self.pathway)

        displacement = self.pathway.length - self.l
        duffing_force = -self.k1 * displacement - self.k2 * displacement**3

        loads = [(self.P1, duffing_force * N.y + damper.force * N.y), (self.P2, -self.m * self.g * N.y)]
        return loads
```

tjstienstra replied:
Why use an entire class for it. You could simplify it to something like:
```python
path = me.LinearPathway(O, block_center)
spring = me.DuffingSpring(k1, k2, path)
damper = me.LinearDamper(c1, path)

loads = spring.to_loads() + damper.to_loads()
```
The gravitational force (potential energy) should not be inside the actuator. To set the potential energy you can do something like:
```python
for body in bodies:
    body.potential_energy = body.mass * g * body.masscenter.pos_from(O).dot(-N.y)
```

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+================
+
+   >>> M, m, l, k1, k2, c1, g, h, w, d, r = sm.symbols('M, m, l, k1, k2, c1, g, h, w, d, r')
+   >>> q1, q2 = me.dynamicsymbols('q1 q2')
+   >>> q1d, q2d = me.dynamicsymbols('q1 q2', 1)
+   >>> u1, u2 = me.dynamicsymbols('u1 u2')
+   >>> u1d, u2d = me.dynamicsymbols('u1 u2', 1)
```
```suggestion
   >>> q1, q2, u2, u2 = me.dynamicsymbols('q1 q2 u1 u2')
```
You don't use time-derivatives manually, so you can remove them.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+Duffing Oscillator with a Pendulum
+==================================
+
+In this example we demonstrate the use of functionality provided in
+:obj:`sympy.physics.mechanics` for deriving the quations of motion (EOM) for a system
+consisting of a Duffing oscillator with a pendulum.
```
Would be nice to refer that this example is inspired by that paper (I think you can use something like `:cite:...`).

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+- :math:`r`: Radius of the massive bob of the pendulum
+- :math:`q_1`: Generalized coordinate representing the position of the Duffing oscillator
+- :math:`q_2`: Generalized coordinate representing the angle of the pendulum
+- :math:`\dot{q}_1`: First time derivative of `q1`, representing the velocity of the Duffing oscillator
+- :math:`\dot{q}_2`: First time derivative of `q2`, representing the angular velocity of the pendulum
```
These can be removed, they are self-explanatory.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+- :math:`\dot{q}_2`: First time derivative of `q2`, representing the angular velocity of the pendulum
+- :math:`u_1`: Generalized speed associated with the Duffing oscillator
+- :math:`u_2`: Generalized speed associated with the pendulum
+- :math:`\dot{u}_1`: First time derivative of `u1`, representing the acceleration of the Duffing oscillator
+- :math:`\dot{u}_2`: First time derivative of `u2`, representing the angular acceleration of the pendulum
```
These can be removed, they are self-explanatory.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+
+The angular velocity of the pendulum in the reference frame is:
+
+   >>> B.set_ang_vel(N, q2.diff() * N.z)
```
This angular velocity has already been set automatically when orieting the frame, so you can remove this line.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+   >>> Pendulum.v2pt_theory(Block, N, B)
+   Derivative(q1(t), t)*N.y - l*Derivative(q2(t), t)*B.x
+
+   >>> me.init_vprinting()
```
You should move this line just after the imports, e.g.:
```
>>> import sympy as sm
>>> import sympy.physics.mechanics as me
>>> me.init_vprinting()
```

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+
+   >>> me.init_vprinting()
+
+Define inertia and rigid bodies.
+Here, we assume a simple pendulum which consists of a bob of mass m hanging from a massless string of length l
+and fixed at a pivot point (Duffing Oscillator Block).
+
+   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
+   >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)
+
+   >>> par_block = me.RigidBody('block', Block, N, M, (I_block, Block))
+   >>> par_pendulum = me.RigidBody('pendulum', Pendulum, B, m, (I_pendulum, Pendulum))
```
I would propose just using particles as done in the paper. This allows comparing the final equations to the paper more easily and it makes the equations more readable as they are simpler.

When doing so, don't forget to remove `h`, `d`, `w`, and `r` from the start of the example.

eh111eh replied:
I compared the equations of motion derived from `LagrangesMethod `.
Equations of motion in the paper:
$$(M+m)q_{1}'' - mlq_{2}''sin(q_{2})-ml(q_{2}')^2cos(q_{2})+k_{1}q_{1}+k_{2}(q_{1})^3+c_{1}q_{1}'$$
$$ml^2q_{2}''-mlq_{1}''sin(q_{2})+mlgsin(q_{2})+c_{2}q_{2}'$$
Equations of motion in this example:
$$(M+m)q_{1}'' - mlq_{2}''sin(q_{2})-ml(q_{2}')^2cos(q_{2})+k_{1}q_{1}+k_{2}(q_{1})^3+c_{1}q_{1}' -(M+m)g$$
$$ml^2q_{2}''-mlq_{1}''sin(q_{2})+mlgsin(q_{2})+ \frac{2r^2q_{2}''}{5}$$

The differences come from the gravitational force and the existence of the damping torque with damping coefficient $c_2$, so I believe equations of motion obtained in this example are acceptable

tjstienstra replied:
Part of the reason was that it also kept the tutorial simpler (the block doesn't rotate so those inertia terms are by definition not showing up in any of the equations), but keeping it is fine with me.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+   >>> bodies = [par_block, par_pendulum]
+
+   >>> for body in bodies:
+   ...     gravitational_force = -body.mass * g * N.y
+   ...     gravitational_load = (body.masscenter, gravitational_force)
+   ...     loads.append(gravitational_load)
```
```suggestion
   ...     loads.append(me.Force(body, -body.mass * g * N.y))
```

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+    --------- + ------------ + ----------------------------------------------------
+        2            5                                  2
+
+   >>> LM = me.LagrangesMethod(L, [q1, q2], forcelist = loads, frame = N)
```
```suggestion
   >>> LM = me.LagrangesMethod(L, [q1, q2], bodies=bodies, forcelist=loads, frame=N)
```
Though `LangrangesMethod` doesn't make use of the bodies, it is nice to provide them anyways. Additionally, when calling a function keyword arguments are generally specified without spacing.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+        2            5                                  2
+
+   >>> LM = me.LagrangesMethod(L, [q1, q2], forcelist = loads, frame = N)
+   >>> LM.form_lagranges_equations()
```
```suggestion
   >>> sm.simplify(LM.form_lagranges_equations())
```
Would recommend comparing results of `sm.simplify` and something like `sm.expand` to see which is more readable and looks closer to the EoMs shown in the paper (they are of course slightly different as you describe a slightly different system).

tjstienstra reviewed (changes requested):
Nice example, which is mostly finished. I left a few comments to further improve it. Optionally, you can also write a short paragraph at the end comparing the equations to the ones in the paper if you like.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,141 @@
+Locations and velocities of the Duffing Oscillator block and the pendulum are:
+
+   >>> O = me.Point('O')
+   >>> Block = O.locatenew('Block', q1 * N.y)
+   >>> Pendulum = Block.locatenew('Pendulum', l * B.y)
```
```suggestion
   >>> block = O.locatenew('block', q1 * N.y)
   >>> pendulum = block.locatenew('pendulum', l * B.y)
```
Use lowercase, according to the pep8 standard as mentioned by @moorepants

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into duffingexample2

eh111eh committed:
Merge branch 'duffingexample2' of https://github.com/eh111eh/SymPy into duffingexample2

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Merge branch 'duffingexample2' of https://github.com/eh111eh/SymPy into duffingexample2

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

eh111eh committed:
Update duffing-example.rst

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+==================================
+
+In this example we demonstrate the use of functionality provided in
+:obj:`sympy.physics.mechanics` for deriving the quations of motion (EOM) for a system
```
```suggestion
:obj:`sympy.physics.mechanics` for deriving the equations of motion for a system
```
You are not using the abbreviation anywhere in this tutorial.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-oscillator-pendulum.svg:
<img src="https://github.com/sympy/sympy/assets/97806294/6e08148a-29a0-42ff-a605-38dd1404a4a4" alt="reference frame" width="150"/>

Can you change the reference frame to exactly match the example. In you current drawing Nz would be pointing inward.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+Here, we assume a simple pendulum which consists of a bob of mass m hanging from a massless string of length l
+and fixed at a pivot point (Duffing Oscillator Block).
+
+   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
```
```suggestion
   >>> I_block = M / 12 * me.inertia(N, h**2 + d**2, w**2 + d**2, w**2 + h**2)
```
Inertia was incorrect. You could optionally use the `Inertia` object to immediately also specify the point w.r.t. which the inertia is defined.
```suggestion
   >>> I_block = me.Inertia.from_inertia_scalars(bock_point, N, M*(h**2 + d**2)/12, M*(w**2 + d**2)/12, M*(w**2 + h**2)/12)
```

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+and fixed at a pivot point (Duffing Oscillator Block).
+
+   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
+   >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)
```
```suggestion
   >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 1, 1)
```
Same here.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
+   >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)
+
+   >>> par_block = me.RigidBody('block', block_point, N, M, (I_block, block_point))
+   >>> par_pendulum = me.RigidBody('pendulum', pendulum_point, B, m, (I_pendulum, pendulum_point))
```
Maybe rename to `block_body` or `block_rb` and `pendulum_...`, as you are modeling them as rigid bodies and not as particles.

tjstienstra commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. math::
+
+   (M + m)q_1'' - mlq_2'' \sin(q_2) - ml(q_2')^2 \cos(q_2) + k_1 q_1 + k_2 (q_1)^3 + c_1 q_1' = F_0 \cos(\nu t)
+
+   ml^2 q_2'' - mlq_1'' \sin(q_2) + mlg \sin(q_2) + c_2 q_2' = 0
+
```
Isn't it possible to use `\ddot{q}_i` and `\dot{q}_i` here?

tjstienstra replied:
Additionally, you can probably replace `(q_1)^3` by `q_1^3`

tjstienstra reviewed (changes requested):
Thank you for the quick adjustments. I have a few follow up comments, but would say it's almost there.

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Timo Stienstra <97806294+tjstienstra@users.noreply.github.com>

moorepants commented:
I see that the doctests in CI are failing. If you don't know yet, you should run `bin/doctest path/to/your/rst/file.rst` to run it locally and debug.

eh111eh committed:
Update duffing-example.rst

All feedback applied
- Created an `mechanics/index.rst`
- Changed the reference frame of schematic
- Renamed par_block and par_pendulum into body_block and body_pendulum
- Changed eom in paper: q1, q2 into y and phi

eh111eh committed:
Merge branch 'duffingexample2' of https://github.com/eh111eh/SymPy into duffingexample2

eh111eh committed:
Update index.rst

moorepants commented:
The SVG has the dark mode issue:
![image](https://github.com/sympy/sympy/assets/276007/9eba4cda-f72e-4014-aaec-c7a37d5b7f4e)

@tjstienstra didn't you have a way to fix this?

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+- :math:`q_1`: Generalized coordinate representing the position of the Duffing oscillator
+- :math:`q_2`: Generalized coordinate representing the angle of the pendulum
+- :math:`u_1`: Generalized speed associated with the Duffing oscillator
+- :math:`u_2`: Generalized speed associated with the pendulum
```
change the u variables to qdots, right?

eh111eh replied:
Yeah, I'll remove u variables

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+With the problem setup, the Lagrangian can be calculated, and the equations of motion formed.
+
+   >>> L = me.Lagrangian(N, block_body, pendulum_body)
+   >>> me.Lagrangian(N, block_body, pendulum_body)
```
```suggestion
   >>> L
```

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+
+With the problem setup, the Lagrangian can be calculated, and the equations of motion formed.
+
+   >>> L = me.Lagrangian(N, block_body, pendulum_body)
```
The Lagrangian should technically include potential energy term. For this model, it is gravity and the Duffing spring. You can add them as loads, but it is probably worth stating that you are setting the potential energy to zero in the Lagrangian and including the conservative forces in the loads, which is equivalent.

moorepants commented:
I left a couple of minor comments, but otherwise this look good to me and I give +1 to merge.

End the end it does expose how the Actuator class is designed to work more with Kane's method where all forces are lumped together and there is no separation of conservative and non-conservative forces like in Lagrange's method. I suppose Actuator could have a potential energy method and then somehow also return the forces minus the potential energy ones. Then the parts needed for forming Lagrange's equations can also be extracted from Actuator. We could open an issue about it.

tjstienstra commented:
> The SVG has the dark mode issue:
> ![image](https://github.com/sympy/sympy/assets/276007/9eba4cda-f72e-4014-aaec-c7a37d5b7f4e)
> 
> @tjstienstra didn't you have a way to fix this?

#25702 is the PR where I fixed it by manually updating the SVGs with size specification and `"CurrentColor"` in the SVGs and `.. raw:: html` in the rst files.

eh111eh committed:
Update duffing-example.rst and svg file

Removed u variables, mentioned usage of conservative forces in the loads, and fixed svg dark mode issue

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+and fixed at a pivot point (Duffing Oscillator Block).
+
+   >>> I_block = M / 12 * me.inertia(N, h**2 + d**2, w**2 + d**2, w**2 + h**2)
+   >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 1, 1)
```
```suggestion
   >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 0, 1)
```

moorepants commented on doc/src/tutorials/physics/mechanics/duffing-example.rst:
```diff
@@ -0,0 +1,146 @@
+
+   ml^2 q_2'' - mlq_1'' \sin(q_2) + mlg \sin(q_2) + \frac{2r^2q_2''}{5} = 0
+
+The differences in the equations of motion are attributed to several factors: the gravitational force, a damping torque characterized by the damping coefficient `c_2`,
```
I don't see $c_2$ in the equations. Is this a typo?

eh111eh replied:
$c_2$ appears in the equations of motion in the paper but not in our example because the paper introduces a damping torque, but our example does not. So it is normal that $c_2$ does not appear in line 136.

moorepants replied:
Ah, I read the equation and then you mention c_2 so I was wondering what that is. But I hadn't read further to the next equation.

moorepants replied:
Actually, that is wrong. I still don't see a c_2 in either of the equations you write.

moorepants replied:
Nevermind. I'm blind.

moorepants commented:
Last two comments from me and then I think it is all correct.

eh111eh committed:
Update doc/src/tutorials/physics/mechanics/duffing-example.rst

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

moorepants commented:
Ok, this is in! Thanks for correcting all the feedback.
